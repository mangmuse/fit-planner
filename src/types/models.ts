import { expectType } from "tsd";
import { User, Exercise, WorkoutDetail, Workout } from "@prisma/client";
import { boolean, nullable, z } from "zod";

export const clientWorkoutSchema = z.object({
  id: z.string(),
  userId: z.string(),
  date: z.string(),
  status: z.enum(["EMPTY", "PLANNED", "IN_PROGRESS", "COMPLETED"]),
  createdAt: z.string(),
  updatedAt: z.string().nullable().optional(),
});

export const clientExerciseSchema = z.object({
  id: z.number(),
  category: z.string(),
  createdAt: z.string(),
  imageUrl: z.string(),
  isBookmarked: z.boolean(),
  isCustom: z.boolean(),
  name: z.string(),
  unit: z.enum(["kg", "lbs"]),
  userId: z.string().nullable(),
  updatedAt: z.string().optional().nullable(),
});

export const clientWorkoutDetailSchema = z.object({
  id: z.string(),
  exerciseId: z.number(),
  exerciseName: z.string(),
  exerciseOrder: z.number(),
  isDone: z.boolean(),
  setOrder: z.number(),
  workoutId: z.string(),
  weight: z.number().nullable(),
  setType: z.enum(["NORMAL", "WARMUP", "DROP", "FAILURE", "AMRAP"]),
  rpe: z.number().nullable(),
  reps: z.number().nullable(),
  createdAt: z.string(),
  updatedAt: z.string().optional().nullable(),
});

export const localWorkoutSchema = z.object({
  serverId: z.string().nullable(),
  updatedAt: z.string().optional().nullable(),
  id: z.number().optional(),
  userId: z.string(),
  createdAt: z.string(),
  date: z.string(),
  isSynced: z.boolean(),
  status: z.enum(["EMPTY", "PLANNED", "IN_PROGRESS", "COMPLETED"]),
});

export const localExerciseSchema = z.object({
  imageUrl: z.string(),
  createdAt: z.string(),
  isCustom: z.boolean(),
  isBookmarked: z.boolean(),
  name: z.string(),
  category: z.string(),
  serverId: z.number().nullable(),
  unit: z.enum(["kg", "lbs"]),
  id: z.number().optional(),
  updatedAt: z.string().nullable().optional(),
  isSynced: z.boolean(),
  userId: z.string().nullable(),
});

export const localWorkoutDetailSchema = z.object({
  id: z.number().optional(),
  serverId: z.string().nullable(),
  workoutId: z.number(),
  weight: z.number().nullable(),
  reps: z.number().nullable(),
  rpe: z.number().nullable(),
  isDone: z.boolean(),
  isSynced: z.boolean(),
  setOrder: z.number(),
  setType: z.enum(["NORMAL", "WARMUP", "DROP", "FAILURE", "AMRAP"]),
  exerciseOrder: z.number(),
  exerciseName: z.string(),
  exerciseId: z.number(),
  createdAt: z.string(),
  updatedAt: z.string().optional().nullable(),
});

// expectType<>({} as LocalExercise);
export type ModelWithStringDates<T> = Omit<T, "createdAt" | "updatedAt"> & {
  createdAt: string;
  updatedAt?: string | null;
};

export type ClientUser = ModelWithStringDates<User>;
export type ClientExercise = z.infer<typeof clientExerciseSchema>;
export type ClientWorkout = z.infer<typeof clientWorkoutSchema>;
export type ClientWorkoutDetail = z.infer<typeof clientWorkoutDetailSchema>;

export type LocalExercise = z.infer<typeof localExerciseSchema>;
export type LocalWorkout = z.infer<typeof localWorkoutSchema>;
export type LocalWorkoutDetail = z.infer<typeof localWorkoutDetailSchema>;
export type LocalWorkoutDetailWithServerWorkoutId = Omit<
  LocalWorkoutDetail,
  "workoutId"
> & {
  workoutId: string;
};

export type AddLocalWorkoutDetailInput = {
  exerciseId: number;
  workoutId: number;
  exerciseOrder: number;
  exerciseName: string;
  setOrder: number;
};
