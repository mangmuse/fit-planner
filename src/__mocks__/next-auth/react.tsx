import React from "react";

export const SessionProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  return <>{children}</>;
};

export const useSession = jest.fn(() => ({
  data: {
    user: { id: "user123", email: "test@example.com", image: "/" },
    expires: "2024-12-31T23:59:59.999Z",
  },
  status: "authenticated",
  update: jest.fn(),
}));

export const signIn = jest.fn();
export const signOut = jest.fn();
export const getSession = jest.fn();
