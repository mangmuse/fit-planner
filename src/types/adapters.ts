import {
  ClientExercise,
  ClientWorkoutDetail,
  LocalExercise,
  LocalRoutine,
  LocalRoutineDetail,
  LocalRoutineDetailWithServerRoutineId,
  LocalWorkout,
  LocalWorkoutDetail,
  LocalWorkoutDetailWithServerWorkoutId,
  NestedExercise,
  Saved,
} from "./models";

export type WD_NewInput = {
  workoutId: number;
  startOrder: number;
};

export type RD_NewInput = {
  routineId: number;
  startOrder: number;
};

// --- SyncAll Adapter --- //

export interface ISyncAllAdapter {
  createNestedStructure: <
    P extends number,
    T extends { id: P },
    K extends PropertyKey,
    U extends Record<K, P>,
  >(
    parents: T[],
    children: U[],
    foreignKey: K
  ) => (T & { details: U[] })[];

  createNestedExercises: (
    exercises: Saved<LocalExercise>[]
  ) => NestedExercise[];
}

export interface IExerciseAdapter {
  mergeServerExerciseData: (
    serverData: ClientExercise[],
    localData: LocalExercise[]
  ) => LocalExercise[];
}

export interface IRoutineDetailAdapter {
  getInitialRoutineDetail(weightUnit?: "kg" | "lbs"): LocalRoutineDetail;

  createRoutineDetail(
    override: Partial<LocalRoutineDetail>,
    weightUnit?: "kg" | "lbs"
  ): LocalRoutineDetail;

  getNewRoutineDetails(
    selectedExercises: { id: number | undefined; name: string }[],
    input: RD_NewInput,
    weightUnit?: "kg" | "lbs"
  ): LocalRoutineDetail[];

  getAddSetToRoutineByLastSet(
    lastSet: LocalRoutineDetail,
    weightUnit?: "kg" | "lbs"
  ): LocalRoutineDetail;

  mapPastWorkoutToRoutineDetail(
    pastWorkoutDetail: LocalWorkoutDetail,
    targetRoutineId: number,
    newExerciseOrder: number,
    weightUnit?: "kg" | "lbs"
  ): LocalRoutineDetail;

  mapLocalRoutineDetailToServer(
    detail: LocalRoutineDetail,
    exercise: LocalExercise,
    routine: LocalRoutine
  ): LocalRoutineDetailWithServerRoutineId;

  cloneToCreateInput(
    input: LocalRoutineDetail,
    newRoutineId: number,
    weightUnit?: "kg" | "lbs"
  ): LocalRoutineDetail;
}

// --- WorkoutDetail Adapter Interface ---
export interface IWorkoutDetailAdapter {
  getInitialWorkoutDetail(weightUnit?: "kg" | "lbs"): LocalWorkoutDetail;
  createWorkoutDetail(
    override: Partial<LocalWorkoutDetail>,
    weightUnit?: "kg" | "lbs"
  ): LocalWorkoutDetail;
  mapPastWorkoutToWorkoutDetail(
    pastWorkoutDetail: LocalWorkoutDetail,
    targetWorkoutId: number,
    newExerciseOrder: number,
    weightUnit?: "kg" | "lbs"
  ): LocalWorkoutDetail;
  getAddSetToWorkoutByLastSet(
    lastSet: LocalWorkoutDetail,
    weightUnit?: "kg" | "lbs"
  ): LocalWorkoutDetail;
  getNewWorkoutDetails(
    selectedExercises: { id: number | undefined; name: string }[],
    input: WD_NewInput,
    weightUnit?: "kg" | "lbs"
  ): LocalWorkoutDetail[];
  mapLocalWorkoutDetailToServer(
    detail: LocalWorkoutDetail,
    relatedExercise: LocalExercise,
    relatedWorkout: LocalWorkout
  ): LocalWorkoutDetailWithServerWorkoutId;
  createOverwriteWorkoutDetailPayload(
    detail: ClientWorkoutDetail,
    exercise: LocalExercise,
    workout: LocalWorkout
  ): LocalWorkoutDetail;
  convertRoutineDetailToWorkoutDetailInput(
    routineDetail: LocalRoutineDetail,
    workoutId: LocalWorkoutDetail["workoutId"],
    weightUnit?: "kg" | "lbs"
  ): LocalWorkoutDetail;
}
