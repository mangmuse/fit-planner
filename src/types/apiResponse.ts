export type ApiErrorResponse = {
  success: false;
  message: string;
};

export type ApiSuccessResponse<T = Record<string, unknown>> = {
  success: true;
} & T;
