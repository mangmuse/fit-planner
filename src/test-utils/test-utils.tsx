import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ModalProvider } from "@/providers/contexts/ModalContext";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import SessionProviderWrapper from "@/providers/SessionProviderWrapper";
import Providers from "@/providers/Providers";

export function customRender(ui: React.ReactElement, options?: RenderOptions) {
  return render(ui, { wrapper: Providers, ...options });
}

export * from "@testing-library/react";
