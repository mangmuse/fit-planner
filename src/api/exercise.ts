import { BASE_URL } from "@/constants";
import { PatchBookmarkInput } from "@/types/dto/exercise.dto";
import { Category, ExerciseType } from "@/types/filters";
import { ClientUser } from "@/types/models";

// export const getAllExercises = async (
//   userId: ClientUser["id"] | undefined,
//   keyword: string,
//   exerciseType: ExerciseType,
//   category: Category
// ) => {
//   const queryParams = new URLSearchParams({
//     userId: userId?.toString() ?? "",
//     keyword: encodeURIComponent(keyword),
//     type: encodeURIComponent(exerciseType),
//     category: encodeURIComponent(category),
//   });

//   const res = await fetch(`${BASE_URL}/api/exercises/all?${queryParams}`);
//   if (!res.ok) {
//     throw new Error("Failed to fetch exercises");
//   }
//   return res.json();
// };
export const getAllExercises = async (userId: ClientUser["id"] | undefined) => {
  const queryParams = new URLSearchParams({
    userId: userId?.toString() ?? "",
  });

  const res = await fetch(`${BASE_URL}/api/exercises/all?${queryParams}`);
  if (!res.ok) {
    throw new Error("Failed to fetch exercises");
  }
  return res.json();
};

export const patchBookmark = async (patchBookmarkInput: PatchBookmarkInput) => {
  const res = await fetch(`${BASE_URL}/api/exercises/bookmark`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(patchBookmarkInput),
  });
  if (!res.ok) {
    throw new Error("Failed to fetch exercises");
  }
  return res.json();
};
