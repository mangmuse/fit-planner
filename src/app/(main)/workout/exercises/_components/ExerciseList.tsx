import ExerciseItem from "@/app/(main)/workout/exercises/_components/ExerciseItem";
import { ClientExerise } from "@/types/models";

type ExerciseListProps = {
  exercises: ClientExerise[];
};
const ExerciseList = ({ exercises }: ExerciseListProps) => {
  return (
    <ul className="h-full flex flex-col gap-1 mt-[14px]">
      {exercises.map((exercise: ClientExerise) => (
        <ExerciseItem key={exercise.id} exercise={exercise} />
      ))}
    </ul>
  );
};

export default ExerciseList;
