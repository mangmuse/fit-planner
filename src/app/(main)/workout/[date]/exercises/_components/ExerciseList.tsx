import ExerciseItem from "@/app/(main)/workout/[date]/exercises/_components/ExerciseItem";
import { ExerciseQueryParams } from "@/types/dto/exercise.dto";
import { ClientExercise, ClientUser, LocalExercise } from "@/types/models";

type ExerciseListProps = {
  exercises: LocalExercise[];
  selectedExercises: { id: number; name: string }[];
  onAdd: (newExercise: LocalExercise) => void;
  onDelete: (toBeDeleted: ClientExercise["id"]) => void;
  onReload: () => void;
  userId: ClientUser["id"];
};
const ExerciseList = ({
  exercises,
  onAdd,
  onDelete,
  onReload,
  selectedExercises,
  userId,
}: ExerciseListProps) => {
  return (
    <ul className="h-full flex flex-col gap-1 mt-[14px]">
      {exercises.map((exercise: LocalExercise) => (
        <ExerciseItem
          key={exercise.id}
          isSelected={selectedExercises.some(
            (selected) => selected.id === exercise.id
          )}
          onAdd={onAdd}
          onDelete={onDelete}
          exercise={exercise}
          onReload={onReload}
        />
      ))}
    </ul>
  );
};

export default ExerciseList;
