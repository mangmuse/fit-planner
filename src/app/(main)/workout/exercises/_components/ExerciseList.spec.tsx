import { mockLocalExercises } from "@/__mocks__/exercise.mock";
import ExerciseList from "@/app/(main)/workout/exercises/_components/ExerciseList";
import { customRender } from "@/test-utils/test-utils";
import { LocalExercise } from "@/types/models";
import { screen } from "@testing-library/react";

describe("ExerciseList", () => {
  const renderList = (exercises: LocalExercise[]) => {
    const selectedExercises = [
      { id: 1, name: "벤치프레스" },
      { id: 2, name: "플랭크" },
    ];

    return customRender(
      <ExerciseList
        exercises={exercises}
        selectedExercises={selectedExercises}
        onAdd={jest.fn()}
        onDelete={jest.fn()}
        onReload={jest.fn()}
      />
    );
  };
  it("exercises 배열의 길이만큼 올바르게 ExerciseItem 컴포넌트를 렌더링한다", () => {
    const { getByText } = renderList(mockLocalExercises);
    mockLocalExercises.forEach((ex) => {
      const item = getByText(ex.name);
      expect(item).toBeInTheDocument();
    });
  });
  it("exercises 이 빈 배열인 경우 ExerciseItem 컴포넌트를 렌더링하지 않는다", () => {
    const { queryByRole } = renderList([]);
    expect(queryByRole("listitem")).not.toBeInTheDocument();
  });

  describe("snapshot", () => {
    it("렌더링 결과가 스냅샷과 일치하는지 확인한다", () => {
      const { asFragment } = renderList(mockLocalExercises);
      expect(asFragment()).toMatchSnapshot();
    });
    it("빈 exercises 배열일 때, 스냅샷 결과가 일치하는지 확인한다", () => {
      const { asFragment } = renderList([]);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
