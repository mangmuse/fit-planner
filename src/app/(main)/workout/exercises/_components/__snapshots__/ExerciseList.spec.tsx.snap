// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseList snapshot 렌더링 결과가 스냅샷과 일치하는지 확인한다 1`] = `
<DocumentFragment>
  <ul
    class="h-full flex flex-col gap-1 mt-[14px]"
  >
    <li
      aria-labelledby="name"
      class="px-3 flex justify-between w-full h-[51px] rounded-lg bg-bg-surface"
    >
      <div
        class="flex items-center gap-5"
      >
        <div
          class="bg-slate-300 w-8 h-8"
        />
        <span
          class="text-primary"
        >
          벤치프레스
        </span>
      </div>
      <div
        class="self-center"
        role="button"
      >
        <img
          alt="북마크 해제"
          src="/img.jpg"
        />
      </div>
    </li>
    <li
      aria-labelledby="name"
      class="px-3 flex justify-between w-full h-[51px] rounded-lg bg-bg-surface"
    >
      <div
        class="flex items-center gap-5"
      >
        <div
          class="bg-slate-300 w-8 h-8"
        />
        <span
          class="text-primary"
        >
          스쿼트
        </span>
      </div>
      <div
        class="self-center"
        role="button"
      >
        <img
          alt="북마크"
          src="/img.jpg"
        />
      </div>
    </li>
    <li
      aria-labelledby="name"
      class="px-3 flex justify-between w-full h-[51px] rounded-lg bg-bg-surface"
    >
      <div
        class="flex items-center gap-5"
      >
        <div
          class="bg-slate-300 w-8 h-8"
        />
        <span
          class=""
        >
          데드리프트
        </span>
      </div>
      <div
        class="self-center"
        role="button"
      >
        <img
          alt="북마크 해제"
          src="/img.jpg"
        />
      </div>
    </li>
    <li
      aria-labelledby="name"
      class="px-3 flex justify-between w-full h-[51px] rounded-lg bg-bg-surface"
    >
      <div
        class="flex items-center gap-5"
      >
        <div
          class="bg-slate-300 w-8 h-8"
        />
        <span
          class=""
        >
          랫풀다운
        </span>
      </div>
      <div
        class="self-center"
        role="button"
      >
        <img
          alt="북마크 해제"
          src="/img.jpg"
        />
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ExerciseList snapshot 빈 exercises 배열일 때, 스냅샷 결과가 일치하는지 확인한다 1`] = `
<DocumentFragment>
  <ul
    class="h-full flex flex-col gap-1 mt-[14px]"
  />
</DocumentFragment>
`;
